### PR TITLE
[WIP][expo-ar] AR universal module for both Android and iOS

### PR DESCRIPTION
**Pull request by bbarthec**
# Why

Resolves https://github.com/expo/universe/issues/2927

# How

- [x] Added ARCore as Android AR library
- [x] Moved AR to module
- [x] Added `expo-three` as package (probably to be moved out)
- [x] Moved whole JS code to TS
- [x] Merged `expo-graphics` with `expo-ar` module. `expo-graphics` would not be needed anymore

# Test Plan

NCL has AR section for AR specific screens.

# Things to still be done:

- [ ] `expo-three` should possibly be not included in this repo
- [ ] iOS refactor & moving camera texture rendering to separate `shared GL Context` (as it's currently done on Android)
- [ ] iOS features that needs to be adapted on Android
- [ ] documentation
- [ ] `expo-three#AR.{MagneticObjects, Light, ShadowFloor, CubeTexture}, loaders & other functions` are not adapted nor tested
- [ ] missing ARKit features 
- [ ] missing ARCore features

### CODE REVIEW

**Comment by bbarthec** [comment]
@tsapeta @sjchmiela @EvanBacon 
Please review this part of whole Android AR introduction.
The rest of functionalities that would came to your mind or are listed above would be made in separate PRs as this one's growing too much. Review with this fought in mind please 😉

File: packages/expo-gl/android/src/main/java/expo/modules/gl/thread/GLSharedThread.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    mEGLDisplay = parentEGLDisplay;
+    mEGLConfig = parentEGLConfig;
+  }
+
+  @Override
+  void initEGL() {
+    mEGL = (EGL10) EGLContext.getEGL();
+
+    mEGLSurface = createEGLSurface(mEGLDisplay, mEGLConfig);
+    if (mEGLSurface == null || mEGLSurface == EGL10.EGL_NO_SURFACE) {
```
Comment (tsapeta) [inline packages/expo-gl/android/src/main/java/expo/modules/gl/thread/GLSharedThread.java]: Is it possible that `mEGLSurface` would be `null` here? I'd recommend to use `EGL_NO_*` flags in all those functions that create EGL objects.

File: packages/expo-gl/android/src/main/java/expo/modules/gl/thread/GLSharedThread.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+            EGL14.eglGetCurrentDisplay(),
+            EGL14.eglGetCurrentSurface(EGL14.EGL_DRAW),
+            EGL14.EGL_SWAP_BEHAVIOR,
+            EGL14.EGL_BUFFER_PRESERVED
+            );
+      }
+    });
+  }
+
+  private EGLSurface createEGLSurface(EGLDisplay eglDisplay, EGLConfig eglConfig) {
```
Comment (tsapeta) [inline packages/expo-gl/android/src/main/java/expo/modules/gl/thread/GLSharedThread.java]: This code already exists in `GLThread.createSurface`. Maybe it should sit in `GLBaseThread` too? Imho everything that is not related to the `shared context` concept shouldn't be here.

File: packages/expo-gl/android/src/main/java/expo/modules/gl/thread/GLSharedThread.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+  @Override
+  void initEGL() {
+    mEGL = (EGL10) EGLContext.getEGL();
+
+    mEGLSurface = createEGLSurface(mEGLDisplay, mEGLConfig);
+    if (mEGLSurface == null || mEGLSurface == EGL10.EGL_NO_SURFACE) {
+      checkEGLError("eglCreateSurface");
+    }
+    mEGLContext = createEGLContext(mEGLDisplay, mEGLConfig, mParentEGLContext, 3);
+    if (mEGLContext == null || mEGLContext == EGL10.EGL_NO_CONTEXT) {
```
Comment (tsapeta) [inline packages/expo-gl/android/src/main/java/expo/modules/gl/thread/GLSharedThread.java]: As above, checking for `null` is probably unnecessary

File: packages/expo-gl/android/src/main/java/expo/modules/gl/thread/GLSharedThread.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+  private EGLContext createEGLContext(EGLDisplay eglDisplay, EGLConfig eglConfig, EGLContext parentContext, int eglContextVersion) {
+    int[] contextAttribs = {
+        EGL_CONTEXT_CLIENT_VERSION, eglContextVersion,
+        EGL10.EGL_NONE,
+    };
+    return mEGL.eglCreateContext(eglDisplay, eglConfig, parentContext, contextAttribs);
+  }
+
+  @Override
+  void deinitEGL() {
```
Comment (tsapeta) [inline packages/expo-gl/android/src/main/java/expo/modules/gl/thread/GLSharedThread.java]: Could you try to merge this with `GLThread.deinitEGL`? 
I guess this should be the default implementation (`GLBaseThread`) and then the thread for the main context should call its superclass method and then call `eglTerminate`.

**Review by tsapeta — CHANGES_REQUESTED** [review]
Left some comments about changes in `expo-gl` part. Looks good 💯, but generally I think the classes for main and shared contexts should be reduced to only what distinguishes them. Does it make sense?

File: apps/native-component-list/package.json [hunk 5]
```diff
@@ -29,7 +29,7 @@
     "react-navigation-header-buttons": "^1.2.1",
     "react-navigation-material-bottom-tabs": "1.0.0-alpha.4",
     "regl": "^1.3.0",
-    "three": "^0.88.0",
+    "three": "^0.98.0",
```
Comment (EvanBacon) [inline apps/native-component-list/package.json]: Make sure that this works with android. If I remember correctly there was some arithmetic error.

File: packages/expo-ar/src/lifecycle/index.ts [hunk 6]
```diff
@@ -0,0 +1,53 @@
+import { findNodeHandle } from 'react-native';
```
Comment (EvanBacon) [inline packages/expo-ar/src/lifecycle/index.ts]: Is there a way to access this without `react-native`?

**Comment by tsapeta** [comment]
Closing due to inactivity 🤷‍♂

**Pull request closed without merging**